File: source_code/aquarium.cpp
#include "aquarium.hpp"
#include "linkedList.hpp"

 // ctor
// Memunculkan satu siput, dimensi akuarium maksimum
aquarium::aquarium(){
}

// dtor
aquarium::~aquarium(){
}

void aquarium::moveAll(double diff, double& money) {
// menggerakkan semua entity yang ada di aquarium
  moveGuppy(diff);
  moveSnail(diff, money);
  movePiranha(diff);
  moveFood(diff);
  moveCoin(diff);
}

void aquarium::moveGuppy(double diff){
// Menggerakkan seluruh guppy
  node<guppy>* tempGuppy = listGuppy.getHead();
  node<guppy>* tempGuppy_2;
  while (tempGuppy != NULL){
    if (tempGuppy->info.GetFullRate() <= -12){
      // Guppy mati
      tempGuppy_2 = tempGuppy->next;
      listGuppy.remove(tempGuppy->info);
      tempGuppy = tempGuppy_2;
    } else {
      tempGuppy->info.produceCoin(diff,listCoin);
      tempGuppy->info.move(diff,listFood);
      tempGuppy = tempGuppy->next;
    }
  }
}

void aquarium::moveSnail(double diff, double& money){
// Mengerakkan seluruh snail
  node<snail>* tempSnail = listSnail.getHead();
  while (tempSnail != NULL){
     tempSnail->info.move(diff,listCoin,money);
     tempSnail = tempSnail->next;
  }
}

void aquarium::movePiranha(double diff){
// Menggerakkan seluruh piranha
  node<piranha>* tempPiranha = listPiranha.getHead();
  node<piranha>* tempPiranha_2;
  while (tempPiranha != NULL){
    if (tempPiranha->info.GetFullRate() <= -12){
      // Piranha mati
      tempPiranha_2=tempPiranha->next;
      listPiranha.remove(tempPiranha->info);
      tempPiranha=tempPiranha_2;
    } else {
      tempPiranha->info.move(diff,listGuppy,listCoin);
      tempPiranha=tempPiranha->next;
    }
  }
}

void aquarium::moveFood(double diff){
// Menggerakkan seluruh food
  node<food>* tempFood = listFood.getHead();
  while (tempFood != NULL){
     tempFood->info.move(diff, listFood);
     tempFood = tempFood->next;
  }
}

void aquarium::moveCoin(double diff){
// Menggerakkan seluruh coin
  node<coin>* tempCoin = listCoin.getHead();
  while(tempCoin != NULL){
    tempCoin->info.move(diff);
    tempCoin = tempCoin->next;
  }
}

void aquarium::addGuppy(){
// menambahkan seekor guppy pada list Guppy
    guppy new_guppy;
    listGuppy.add(new_guppy);
}

void aquarium::addPiranha(){
// menambahkan seekor piranha pada listFish
    piranha new_piranha;
    listPiranha.add(new_piranha);
}

void aquarium::addCoin(){
// menambahkan sebuah koin pada listCoin
    coin new_coin;
    listCoin.add(new_coin);
}

void aquarium::addSnail(){
// Menambahkan seekor siput pada listSnail
    snail new_snail;
    listSnail.add(new_snail);
}

void aquarium::addFood(int x){
  // menambahkan sebuah makanan pada listFood
    food new_food(x);
    listFood.add(new_food);
}

bool aquarium::isThereIsNoFish() {
  return (listGuppy.getHead() == NULL) && (listPiranha.getHead() == NULL);
}

File: Java DOG/aquarium.java
public class aquarium{
    private static final int MAX_LENGTH = 640;
    private static final int MAX_WIDTH = 480;
    private int money;
    private LinkedList<guppy> listGuppy;
    private LinkedList<piranha> listPiranha;
    private LinkedList<coin> listCoin;
    private LinkedList<snail> listSnail;
    private LinkedList<food> listFood;
    public aquarium(){
        listGuppy = new LinkedList<guppy>();
        listPiranha = new LinkedList<piranha>();
        listCoin = new LinkedList<coin>();
        listSnail = new LinkedList<snail>();
        listFood = new LinkedList<food>();
        money = 10000;
    }
    public void moveAll(double diff){
        // menggerakkan semua entity yang ada di aquarium
        moveGuppy(diff);
        moveSnail(diff);
        movePiranha(diff);
        moveFood(diff);
        moveCoin(diff);
    }
    public void moveGuppy(double diff){
        // Menggerakkan seluruh guppy
        Node<guppy> tempGuppy = listGuppy.getHead();
        Node<guppy> tempGuppy_2;
        while (tempGuppy != null){
            if (tempGuppy.getData().GetFullRate() <= -12){
                // Guppy mati
                tempGuppy_2 = tempGuppy.getNext();
                listGuppy.remove(tempGuppy.getData());
                tempGuppy = tempGuppy_2;
            } else {
                tempGuppy.getData().produceCoin(diff,listCoin);
                tempGuppy.getData().move(diff,listFood);
                tempGuppy = tempGuppy.getNext();
            }
        }
    }
        
    public void moveSnail(double diff){
        // Mengerakkan seluruh snail
        Node<snail> tempSnail = listSnail.getHead();
        while (tempSnail != null){
            tempSnail.getData().move(diff,listCoin,money);
            tempSnail = tempSnail.getNext();
        }
    }
    public void movePiranha(double diff){
        // Menggerakkan seluruh piranha
        Node<piranha> tempPiranha = listPiranha.getHead();
        Node<piranha> tempPiranha_2;
        while (tempPiranha != null){
            if (tempPiranha.getData().GetFullRate() <= -12){
                // Piranha mati
                tempPiranha_2=tempPiranha.getNext();
                listPiranha.remove(tempPiranha.getData());
                tempPiranha=tempPiranha_2;
            } else {
                tempPiranha.getData().move(diff,listGuppy,listCoin);
                tempPiranha=tempPiranha.getNext();
            }
        }
    }
        
    public void moveFood(double diff){
        // Menggerakkan seluruh food
        Node<food> tempFood = listFood.getHead();
        while (tempFood != null){
            tempFood.getData().move(diff, listFood);
            tempFood = tempFood.getNext();
        }
    }
        
    public void moveCoin(double diff){
        // Menggerakkan seluruh coin
        Node<coin> tempCoin = listCoin.getHead();
        while(tempCoin != null){
            tempCoin.getData().move(diff);
            tempCoin = tempCoin.getNext();
        }
    }
        
    public void addGuppy(){
        // menambahkan seekor guppy pada list Guppy
        guppy new_guppy = new guppy();
        listGuppy.add(new_guppy);
    }
        
    public void addPiranha(){
        // menambahkan seekor piranha pada listFish
        piranha new_piranha = new piranha();
        listPiranha.add(new_piranha);
    }
        
    public void addCoin(){
        // menambahkan sebuah koin pada listCoin
        coin new_coin = new coin();
        listCoin.add(new_coin);
    }
        
    public void addSnail(){
        // Menambahkan seekor siput pada listSnail
        snail new_snail = new snail();
        listSnail.add(new_snail);
    }
        
    public void addFood(int x){
        // menambahkan sebuah makanan pada listFood
        food new_food = new food(x);
        listFood.add(new_food);
    }
        
    public Boolean isThereIsNoFish() {
        return (listGuppy.getHead() == null) && (listPiranha.getHead() == null);
    }
    public int getMoney(){
        return money;
    }
    public void setMoney(int money){
        this.money = money;
    }
}